src/hypercorn/protocol/http_stream.py

```diff
@@ -181,13 +195,43 @@ async def app_send(self, message: Optional[ASGISendEvent]) -> None:
                     )
 
                 if not message.get("more_body", False):
-                    if self.state != ASGIHTTPState.CLOSED:
+                    await self.send(EndBody(stream_id=self.stream_id))
+
+                    if not self.response.get("trailers", False):
                         self.state = ASGIHTTPState.CLOSED
                         await self.config.log.access(
                             self.scope, self.response, time() - self.start_time
                         )
-                        await self.send(EndBody(stream_id=self.stream_id))
                         await self.send(StreamClosed(stream_id=self.stream_id))
+            elif (
+                message["type"] == "http.response.trailers"
+                and self.scope["http_version"] in TRAILERS_VERSIONS
+                and self.state
+                in {
+                    ASGIHTTPState.REQUEST,
```

Comment:
@jeffsawatzky why allow trailers to be sent before the response?

Reply:
@pgjones because there may not be a response. gRPC uses the trailing headers for error responses, so if there was an error server side before any response was sent then it would just return a trailing header and no response.

Reply:
@pgjones if you look at this:
https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
The response could be trailers only.

Reply:
I think the status would have to be 200 (I don't think there is a `self.response` to use), does this make sense to you?

Reply:
For my usecase with gRPC, yes that makes sense. All gRPC responses are supposed to be 200.
https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses

Whether it makes sense generally, I think so?